sumpy/p2e.py

```diff
@@ -80,24 +81,40 @@ def get_loopy_instructions(self):
         from sumpy.assignment_collection import SymbolicAssignmentCollection
         sac = SymbolicAssignmentCollection()
 
-        coeff_names = [
-            sac.assign_unique("coeff%d" % i, coeff_i)
+        coeff_names = []
+        code_transformers = [self.expansion.get_code_transformer()]
+        for knl_idx, kernel in enumerate(self.kernels):
+            code_transformers.append(kernel.get_code_transformer())
             for i, coeff_i in enumerate(
-                self.expansion.coefficients_from_source(avec, None, rscale, sac))]
+                self.expansion.coefficients_from_source(avec, None, rscale,
+                     sac, kernel=kernel)
+            ):
+                sac.add_assignment(f"coeff{i}_{knl_idx}", coeff_i)
+                coeff_names.append(f"coeff{i}_{knl_idx}")
 
         sac.run_global_cse()
 
         from sumpy.codegen import to_loopy_insns
         return to_loopy_insns(
                 sac.assignments.items(),
                 vector_names={"a"},
-                pymbolic_expr_maps=[self.expansion.get_code_transformer()],
+                pymbolic_expr_maps=code_transformers,
```

Comment:
Just use a separate comprehension here to gather the code transformers. No need to mash expansion and other logic together.